debezium/debezium
# DBZ-4783: Support for multiple databases and tasks in the SQL Server connector

Change summary:
1. Manage change tables for each partition individually.
2. Expose the actual task id to the metrics in order to be able to run multiple tasks.
3. Run the connector with multiple databases and multiple tasks.

Note, running multiple tasks is _not_ necessary for capturing multiple databases. Furthermore, running multiple tasks cannot be tested in the current test suite. The only reason why it's contributed here is that extracting the support for multiple tasks in a separate patch would require extra work (it was originally implemented like this). I can remove it from the patch, if necessary.

**TODO**:
- [x] Provide an integration test for the multi-partition scenario.

### Manual testing of the multi-database and multi-task configuration

1. Modify `docker-compose-sqlserver.yaml` from the [tutorial](https://github.com/debezium/debezium-examples/tree/main/tutorial#using-sql-server) to use the Docker image built from this PR.
2. Start the services:
   ```shell
   export DEBEZIUM_VERSION=1.9
   docker-compose -f docker-compose-sqlserver.yaml up
   ```
3. Initialize test databases:
   ```shell
   docker-compose -f docker-compose-sqlserver.yaml exec -T sqlserver bash -c '/opt/mssql-tools/bin/sqlcmd -U sa -P $SA_PASSWORD' << 'EOF'
   CREATE DATABASE testDB1;
   GO
   USE testDB1;
   EXEC sys.sp_cdc_enable_db;
   
   CREATE TABLE products (
     id INTEGER IDENTITY(101,1) NOT NULL PRIMARY KEY,
     name VARCHAR(255) NOT NULL,
     description VARCHAR(512),
     weight FLOAT
   );
   INSERT INTO products(name,description,weight)
     VALUES ('scooter','Small 2-wheel scooter',3.14);
   
   EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'products', @role_name = NULL, @supports_net_changes = 0;
   GO
   
   CREATE DATABASE testDB2;
   GO
   USE testDB2;
   EXEC sys.sp_cdc_enable_db;
   
   CREATE TABLE customers (
     id INTEGER IDENTITY(1001,1) NOT NULL PRIMARY KEY,
     first_name VARCHAR(255) NOT NULL,
     last_name VARCHAR(255) NOT NULL,
     email VARCHAR(255) NOT NULL UNIQUE
   );
   INSERT INTO customers(first_name,last_name,email)
     VALUES ('Sally','Thomas','sally.thomas@acme.com');
   
   EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'customers', @role_name = NULL, @supports_net_changes = 0;
   GO
   EOF
   ```

4. Start the connector:

   ```shell
   curl -i -X POST -H "Accept:application/json" -H "Content-Type:application/json" http://localhost:8083/connectors/ -d@- << 'EOF'
   {
       "name": "inventory-connector",
       "config": {
           "connector.class" : "io.debezium.connector.sqlserver.SqlServerConnector",
           "tasks.max" : "2",
           "database.server.name" : "server1",
           "database.hostname" : "sqlserver",
           "database.port" : "1433",
           "database.user" : "sa",
           "database.password" : "Password!",
           "database.names" : "testDB1,testDB2",
           "database.history.kafka.bootstrap.servers" : "kafka:9092",
           "database.history.kafka.topic": "schema-changes.inventory"
       }
   }
   EOF
   ```

5. Check that the connector and both tasks are running:

   ```shell
   curl -s http://localhost:8083/connectors/inventory-connector/status
   ```

6. Check that each task exposes metrics for its subset of databases:
   ![Multi-partition and multi-task metrics](https://user-images.githubusercontent.com/59683/155426735-5e0b7ce3-df4b-46a9-880d-0086bd00ca3c.png)

7. Run a consumer and confirm that both databases were snapshotted:

   ```shell
   docker-compose -f docker-compose-sqlserver.yaml exec kafka /kafka/bin/kafka-console-consumer.sh \
     --bootstrap-server kafka:9092 \
     --from-beginning \
     --property print.key=true \
     --whitelist 'server1\.(testDB\d+)\..*'
   ```

8. Make changes in both databases and confirm that the changes from both databases are captured:

   ```shell
   docker-compose -f docker-compose-sqlserver.yaml exec -T sqlserver bash -c '/opt/mssql-tools/bin/sqlcmd -U sa -P $SA_PASSWORD' << 'EOF'
   UPDATE testDB1.dbo.products SET weight = 3.15 WHERE id = 101;
   UPDATE testDB2.dbo.customers SET first_name = 'Molly' WHERE id = 1001;
   EOF
   ```
